avionics/common/include/radio.h

```diff
@@ -69,84 +61,68 @@ class RadioController {
         bulk_sensor = 0x30,
     };
 
+    constexpr static Hal::ms WATCHDOG_SEND_INTERVAL{3000};
+    constexpr static uint16_t STOP_PARSE_FLAG = 1 << 0;
+    constexpr static uint16_t CAN_SEND_FLAG = 1 << 1;
+
   public:
+    // type RadioMembers is public for technical reasons but cannot be used
+    // (especially since no definition is given)
+    class RadioMembers;
+
+    // Loose type to more explicitly indicate what variables are commands.
+    typedef uint8_t command_t;
+
     /**
-     * @brief Constructor.
-     * @param SERIAL_RADIO Uninitialized HardwareSerial used for radio (e.g.
+     * @brief Initializes the class.
+     * @param SERIAL_RADIO Uninitialized HardwareSerial used for radio
+     (e.g.
      * SerialRadio)
-     * @param MAX_QUEUED_BYTES Maximum number of subpacket data bytes to queue
+     * @param MAX_QUEUED_BYTES Maximum number of subpacket data bytes to
+     queue
      * before dropping the oldest subpackets.
-     * @param MAX_PACKETS_PER_RX_LOOP Maximum number of packets to send every
-     * time listenAndAct() is called.
+     * @param MAX_PACKETS_PER_RX_LOOP Maximum number of packets to send
+     every
+     * time forwardCommand() is called.
      */
-    RadioController(Hal::Serial &serial_radio,
-                    unsigned short const MAX_QUEUED_BYTES = 800,
-                    uint8_t const MAX_PACKETS_PER_RX_LOOP = 8)
-        : gnd_addr_(XBeeAddress64(GND_STN_ADDR_MSB, GND_STN_ADDR_LSB)),
-          tx_q_(MAX_QUEUED_BYTES),
-          MAX_PACKETS_PER_RX_LOOP_(MAX_PACKETS_PER_RX_LOOP),
-          WATCHDOG_SEND_INTERVAL_(3000), watchdog_last_send_(Hal::now_ms()) {
-
-        serial_radio.begin(RADIO_BAUD_RATE);
-        while (!serial_radio)
-            ;
-        xbee_.setSerial(serial_radio.getSerial());
-        tx_packet_.setAddress64(gnd_addr_);
-        tx_packet_.setPayload(payload_);
-        sendMessage(Hal::millis(), "Radio initialized");
-        send();
-    }
+    static void initialize();
 
     /**
      * @brief Meat of the action - listens for any incoming packets, then
      * transmits data and performs rocket actions as necessary.
      * @param command_receiver Object with method run_command that
      * accepts an unsigned char (representing the command) and returns void.
      */
-    template <typename Actor> void listenAndAct(Actor &command_receiver) {
+    template <typename Actor>
+    static void forwardCommand(Actor &command_receiver) {
```

Comment:
So now can we move this out of radio now? Radio shouldn't be responsible for executing commands.

Also if you move this into avionics.cpp or rocket.h (or anywhere above radio and rocket) you'd get rid of the circular dependency. Also wouldn't require the template (although I guess we can leave the template in in case we ever want different actors).

P.S:
Oh but I guess you were saying how you didn't want the loop logic to leak out of the radio... Makes sense. I guess neither solution satisfies both. In that case we can leave it as is if you prefer.

Reply:
yep, loop logic should stay in radio. I pulled out a good chunk of the code to a helper function for the purposes of potential unit testing/mocking; the radio.h header should no longer depend on the XBee driver now - just the .cpp file. 

Again - template is to have effectively an interface. Other option is to be more strict and take in only a function pointer - but it's a bit convoluted to make it possible (you have to make the command receiver use global variables, then wrap said global variables in a class with friend access specifier (or use `static`) to limit access, then also juggle the initialization order)